### PR TITLE
feat: dashboard page

### PR DESCRIPTION
**Pull request by tracy-french**
# Description

This change includes the initial IoT App Kit integration. There are plenty of kinks to work out, but we want to move forward with the integration and iterate.

The change also includes a refactor of the app to improve readability and component decomposition. 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] e2e tests passing locally
- [x] build passing locally

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Legal

This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: apps/client/src/hooks/application/use-application.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import { useNavigate } from 'react-router-dom';
+import { useDismissAllNotifications } from '~/hooks/notifications/use-dismiss-all-notifications';
+import { useSetNavigationVisibility } from './use-set-navigation-visibility';
+
+/** Use to control the application. */
+export function useApplication() {
```
Comment (boweihan) [inline apps/client/src/hooks/application/use-application.ts:6]: should this be `useNavigation`? `useApplication` seems maybe a little broad :).

File: apps/client/src/helpers/predicates/is-not-fatal.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { isApiError } from './is-api-error';
+
+import type { ApiError } from '~/services';
+import type { FatalStatusCode } from '~/types/fatal-status-code';
+
+type NotFatalError = ApiError & { status: Exclude<number, FatalStatusCode> };
+
+export function isNotFatal(error: unknown): error is NotFatalError {
+  return isApiError(error) && error.status < 500;
```
Comment (boweihan) [inline apps/client/src/helpers/predicates/is-not-fatal.ts:9]: 👍 we can further segment this later

File: apps/client/src/hooks/application/use-application.ts (hunk 1 above)
Comment (tracy-french) [reply apps/client/src/hooks/application/use-application.ts:6]: Will fix it forward!

File: tests/pages/application-frame.page.ts [hunk 3]
```diff
@@ -23,9 +23,13 @@ export class SideNavigation {
   public readonly dashboardsPageLink: Locator;
 
   constructor(page: Page) {
-    this.navigationDrawer = page.getByRole('navigation', {
-      name: 'Navigation drawer',
-    });
+    this.navigationDrawer = page
+      .getByRole('navigation', {
+        name: 'Navigation drawer', // XXX: Cloudscape gives two elements the same ARIA label
```
Comment (chejimmy) [inline tests/pages/application-frame.page.ts:28]: what does `XXX` mean?

File: apps/client/src/routes/dashboards/dashboard/dashboard-page.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  const dashboardQuery = useDashboardQuery(params.dashboardId);
+  const updateDashboardMutation = useUpdateDashboardMutation();
+
+  if (dashboardQuery.isInitialLoading) {
+    return <DashboardLoadingState />;
+  }
+
+  return (
+    <IoTAppKitDashboard
```
Comment (chejimmy) [inline apps/client/src/routes/dashboards/dashboard/dashboard-page.tsx:24]: where do I find the documentation on IoTAppKitDashboard?

File: apps/client/src/routes/dashboards/dashboard/hooks/use-dashboard-query.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  const emit = useEmitNotification();
+
+  return useQuery({
+    ...createDashboardQuery(dashboardId),
+    useErrorBoundary: isFatal,
+    onError: (error) => {
+      invariant(
+        isNotFatal(error),
+        'Expected fatal error to be handled by error boundary',
+      );
```
Comment (chejimmy) [inline apps/client/src/routes/dashboards/dashboard/hooks/use-dashboard-query.ts:22]: Might not be a good usage of `invariant(...)` here. During fatal error, the error boundary is getting `Error('Invariant violation: ...` and the actual `error` is not passed.

**Review by chejimmy — APPROVED** [review]
LGTM! TY for putting the dashboard in the App 📊🚀

File: apps/client/src/hooks/application/use-application.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  const navigateClient = useNavigate();
+  const dismissNotifications = useDismissAllNotifications();
```
[the remark this answers is not in the document]
Comment (tracy-french) [reply apps/client/src/hooks/application/use-application.ts:8]: https://github.com/awslabs/iot-application/issues/82

File: apps/client/src/hooks/application/use-application.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  const setNavigationVisibility = useSetNavigationVisibility();
+
+  function closeNavigation() {
+    setNavigationVisibility(false);
+  }
+
+  /**
+   * Navigate the client to a new page.
+   *
+   * @regards
```
Comment (square-li) [inline apps/client/src/hooks/application/use-application.ts:18]: I haven't seen this tag before. Is it similar to `@remarks` in tsdoc?

File: apps/client/src/index.tsx [hunk 8]
```diff
@@ -18,9 +18,9 @@ import '@cloudscape-design/global-styles/index.css';
 interface GlobalWithResources extends GlobalThis {
   awsResources: object;
 }
-
 const awsResources = (global as GlobalWithResources).awsResources;
 
+console.log(awsResources);
```
[the remark this answers is not in the document]
Comment (square-li) [reply apps/client/src/index.tsx:23]: +1 we can enforce this by enable `no-console` rule in eslint

https://eslint.org/docs/latest/rules/no-console

File: apps/client/src/routes/dashboards/dashboard/iot-app-kit-dashboard.d.ts [hunk 9]
```diff
@@ -0,0 +1,2 @@
+// TODO: Remove this file when the @iot-app-kit/dashboard declaration file is available.
+declare module '@iot-app-kit/dashboard';
```
Comment (square-li) [inline apps/client/src/routes/dashboards/dashboard/iot-app-kit-dashboard.d.ts:2]: can we remove this? I think dashboard has it declaration at [here](https://github.com/awslabs/iot-app-kit/blob/main/packages/dashboard/package.json#LL22)

File: apps/client/src/routes/dashboards/create/hooks/use-create-dashboard-mutation.tsx [hunk 10]
```diff
@@ -1,37 +1,43 @@
...
 
   return useMutation({
     mutationFn: (formData: Pick<Dashboard, 'name' | 'description'>) => {
       return createDashboard({ ...formData, definition: { widgets: [] } });
     },
-    onMutate: async () => {
+    onMutate: () => {
       // abort any in progress list dashboards queries when mutation starts
-      await queryClient.cancelQueries(DASHBOARD_SUMMARIES_QUERY_KEY);
+      void cancelDashboardsQueries();
```
Comment (square-li) [inline apps/client/src/routes/dashboards/create/hooks/use-create-dashboard-mutation.tsx:32]: I have google this usage of `void` =P

**Review by square-li — APPROVED** [review]

File: apps/client/src/helpers/tests/testing-library.tsx [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+
+function render(ui: ReactElement, { locale = 'en', ...renderOptions } = {}) {
+  function Wrapper(props: PropsWithChildren) {
+    return <IntlProvider locale={locale}>{props.children}</IntlProvider>;
+  }
+  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
+}
+
+// re-export everything
+export * from '@testing-library/react';
```
[the remark this answers is not in the document]
Comment (tracy-french) [reply apps/client/src/helpers/tests/testing-library.tsx:13]: Testing-library is a dev dependency, so it shouldn't be included in the build.

File: apps/client/src/hooks/application/use-application.ts (hunk 7 above)
Comment (tracy-french) [reply apps/client/src/hooks/application/use-application.ts:18]: Oof, it should be `@remarks`!

File: apps/client/src/routes/dashboards/dashboard/iot-app-kit-dashboard.d.ts (hunk 9 above)
Comment (tracy-french) [reply apps/client/src/routes/dashboards/dashboard/iot-app-kit-dashboard.d.ts:2]: @jmbuss fixed this in IoT App Kit and the new release should be out today to resolve.

File: apps/client/src/helpers/predicates/is-fatal.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
+import { isApiError } from './is-api-error';
+import { ApiError } from '~/services';
+
+import type { FatalStatusCode } from '~/types/fatal-status-code';
+
+type FatalError = ApiError & { status: FatalStatusCode };
+
+export function isFatal(error: unknown): error is FatalError {
+  return isApiError(error) && error.status >= 500;
```
[the remark this answers is not in the document]
Comment (chejimmy) [reply apps/client/src/helpers/predicates/is-fatal.ts:9]: I would call `>= 500` retry(i)able
I think any error that would prevent the customer flow to proceed is a fatal error. I think the distinction we should draw here is whether the error can be retried.

**Pull request merged**